### PR TITLE
INT-3860 - Add Repos to Tenable-io Integration & Create Mapped Relationship

### PR DESCRIPTION
**Pull request by eXtremeX**
- refactor Container Security to v2 as v1 has been deprecated.
- the following **new** entities have been added:

| Resources            | Entity `_type`                 | Entity `_class` |
| -------------------- | ------------------------------ | --------------- |
| Container Image      | `tenable_container_image`      | `Image`         |
| Container Repository | `tenable_container_repository` | `Repository`    |
| Service              | `tenable_scanner`              | `Service`       |

### Relationships

- the following **new** relationships have been added:

| Source Entity `_type`          | Relationship `_class` | Target Entity `_type`                |
| ------------------------------ | --------------------- | ------------------------------------ |
| `tenable_account`              | **PROVIDES**          | `tenable_scanner`                    |
| `tenable_account`              | **HAS**               | `tenable_container_image`            |
| `tenable_account`              | **HAS**               | `tenable_container_repository`       |
| `tenable_scanner`              | **SCANS**             | `tenable_container_image`            |
| `tenable_container_image`      | **HAS**               | `tenable_container_report`           |
| `tenable_container_image`      | **HAS**               | `tenable_container_finding`          |
| `tenable_container_image`      | **HAS**               | `tenable_container_malware`          |
| `tenable_container_image`      | **HAS**               | `tenable_container_unwanted_program` |
| `tenable_container_repository` | **HAS**               | `tenable_container_image`            |

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,31 @@ and this project adheres to
...
 ## [Unreleased]
 
+- refactor Container Security to v2 as v1 has been deprecated.
+- the following **new** entities have been added:
+
+| Resources            | Entity `_type`                 | Entity `_class` |
+| -------------------- | ------------------------------ | --------------- |
+| Container Image      | `tenable_container_image`      | `Image`         |
+| Container Repository | `tenable_container_repository` | `Repository`    |
+| Service              | `tenable_scanner`              | `Service`       |
```
Comment (VDubber) [inline CHANGELOG.md:18]: Is `Scanner` not an appropriate class for this entity?

Comment (eXtremeX) [reply CHANGELOG.md:18]: Hey @VDubber, I'm not sure. I've been seeing mentions of `Service` (class: `Service`) more and more in the Jira tickets as well as integrations doing things such as this one: https://github.com/JupiterOne/graph-checkmarx/blob/main/src/constants.ts#L26

We can do whatever is the best of course, just not sure what's the best way to handle these services/scanners in general :thinking:

File: src/steps/service/converters.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    _type: Entities.SERVICE._type,
+    _class: Entities.SERVICE._class,
+    _rawData: [{ name: 'default', rawData: service }],
+    name: service.name,
+    displayName: service.name,
+    category: ['software', 'other'],
+  };
+}
+
+export function createAccountServiceRelationship(
```
Comment (VDubber) [inline src/steps/service/converters.ts:23]: Consider adding tests for this method.

File: src/steps/vulnerabilities/converters.ts [hunk 3]
```diff
@@ -297,16 +296,27 @@ export function createVulnerabilityEntity(
...
         // additional asset properties can be added
         'asset.uuid': vuln.asset.uuid,
         assetHostname: vuln.asset.hostname,
         assetIpv4: vuln.asset.ipv4,
         assetDeviceType: vuln.asset.device_type,
         assetMacAddress: vuln.asset.mac_address,
         agentId: vuln.asset.agent_uuid,
-        first_found: vuln.first_found,
-        last_found: vuln.last_found,
+        firstFound: parseTimePropertyValue(vuln.first_found),
```
Comment (VDubber) [inline src/steps/vulnerabilities/converters.ts:318]: dates should have a suffix of `On`
`firstFound` -> `firstFoundOn`
`lastFound` -> `lastFoundOn`
etc

File: src/steps/vulnerabilities/converters.ts [hunk 4]
```diff
@@ -317,31 +327,27 @@ export function createVulnerabilityEntity(
...
+        exploitabilityEase: vuln.plugin.exploitability_ease,
+        exploitedByMalware: vuln.plugin.exploited_by_malware,
+        exploitedByNessus: vuln.plugin.exploited_by_nessus,
 
         // data model properties
         numericPriority,
         priority,
-        firstSeenOn: getTime(vuln.first_found),
-        lastSeenOn: getTime(vuln.last_found),
+        firstSeenOn: parseTimePropertyValue(vuln.first_found),
```
Comment (VDubber) [inline src/steps/vulnerabilities/converters.ts:349]: ❤️

File: src/steps/vulnerabilities/converters.ts [hunk 5]
```diff
@@ -317,31 +327,27 @@ export function createVulnerabilityEntity(
         riskFactor: vuln.plugin.risk_factor,
         // additional scan properties can be added
         'scan.uuid': vuln.scan.uuid,
-        'scan.started_at': vuln.scan.started_at,
-        'scan.completed_at': vuln.scan.completed_at,
-        severity: vuln.severity,
-        severity_default_id: vuln.severity_default_id,
-        severity_id: vuln.severity_id,
-        severity_modification_type: vuln.severity_modification_type,
+        'scan.startedAt': parseTimePropertyValue(vuln.scan.started_at),
```
Comment (VDubber) [inline src/steps/vulnerabilities/converters.ts:330]: Should we convert this to `'scan.startedOn'`? Not really sure on this one.

File: src/tenable/TenableClient.ts [hunk 6]
```diff
@@ -63,6 +68,19 @@ export default class TenableClient {
       retryMaxAttempts === undefined ? 10 : retryMaxAttempts;
   }
 
+  private async paginated(
```
Comment (VDubber) [inline src/tenable/TenableClient.ts:71]: Please add test for `paginated` in TenableClient.test.ts or move to util and test there.

File: src/tenable/TenableClient.ts [hunk 7]
```diff
@@ -63,6 +68,19 @@ export default class TenableClient {
...
+    cb: (offset: number, limit: number) => Promise<number>,
```
Comment (VDubber) [inline src/tenable/TenableClient.ts:72]: nit: rename cb -> callback

File: src/tenable/TenableClient.ts [hunk 8]
```diff
@@ -330,29 +348,62 @@ export default class TenableClient {
     return { exportUuid };
   }
 
-  public async fetchContainers(): Promise<Container[]> {
-    const containersResponse = await this.retryRequest(() =>
-      this.client.fetchContainers(),
-    );
+  public async iterateContainerRepositories(
```
Comment (VDubber) [inline src/tenable/TenableClient.ts:351]: Please add test coverage in Client file or in the individual step index.ts

**Review by VDubber — CHANGES_REQUESTED** [review]
Looking good! Just a few testing adds and we should be set.

File: src/steps/containers/converters.ts [hunk 9]
```diff
@@ -1,55 +1,150 @@
...
+import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
+import { getServiceKey } from '../service/converters';
 
-export function createContainerEntity(container: Container) {
+function generateImageKey(image: ContainerImage) {
+  return `${image.repoName}:${image.name}:${image.tag}`;
+}
+
+export function createContainerRepositoryEntity(repo: ContainerRepository) {
+  return {
```
[the remark this answers is not in the document]
Comment (eXtremeX) [reply src/steps/containers/converters.ts:26]: Do you want me to refactor just that part over there or everywhere in this repo? Looks like this integration used some old pattern, so we were just trying to add our changes without changing/refactoring all of that. Quite a lot of converters functions aren't using createIntegrationEntity().

File: src/steps/containers/converters.ts [hunk 10]
```diff
@@ -109,22 +207,23 @@ export function createContainerReportRelationship(
...
     _type: Entities.CONTAINER_FINDING._type,
     _class: Entities.CONTAINER_FINDING._class,
     _rawData: [{ name: 'default', rawData: finding }],
+    category: nvdFinding.access_vector,
     displayName: displayName(finding),
     referenceId: nvdFinding.reference_id,
     cve: nvdFinding.cve,
-    publishedDate: nvdFinding.published_date,
-    modifiedDate: nvdFinding.modified_date,
+    publishedDate: parseTimePropertyValue(nvdFinding.published_date),
```
[the remark this answers is not in the document]
Comment (eXtremeX) [reply src/steps/containers/converters.ts:222]: Yep, from the responses we're getting they're timestamps in `string` type.
image

File: src/steps/containers/converters.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (VDubber) [reply src/steps/containers/converters.ts]: My preference is to always add the newest patterns but don't refactor everything. A refactor PR should be a separate task/ticket. It makes the PR easier to review but still pushes good patterns forward.

**Review by VDubber — APPROVED** [review]
Awesome work!

**Pull request merged**